skbio/io/format/genbank.py

```diff
@@ -0,0 +1,468 @@
+            'DBSOURCE',
+            'DBLINK',
+            'KEYWORDS',
+            'SOURCE',
+            'REFERENCE',
+            'COMMENT',
+            'FEATURES',
+            'ORIGIN']
+
+
+@genbank.sniffer()
+def _genbank_sniffer(fh):
+    # check the 1st line starts with 'LOCUS' and the last line ends with '//'
+    if _too_many_blanks(fh, 5):
+        return False, {}
+    try:
+        line = next(_line_generator(fh, skip_blanks=True))
+    except StopIteration:
+        return False, {}
+
+    if line.startswith(_HEADERS[0]):
+        for line in _line_generator(fh, skip_blanks=True):
+            if line == '//':
+                return True, {}
+        return False, {}
+    else:
+        return False, {}
+
+
+@genbank.reader(None)
+def _genbank_to_generator(fh, constructor=None, **kwargs):
+    for record in _parse_genbanks(fh):
+        yield _construct(record, constructor, **kwargs)
+
+
+@genbank.reader(Sequence)
+def _genbank_to_biological_sequence(fh, rec_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), rec_num)
+    return _construct(record, Sequence, **kwargs)
+
+
+@genbank.reader(DNA)
+def _genbank_to_dna(fh, rec_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), rec_num)
+    return _construct(record, DNA, **kwargs)
+
+
+@genbank.reader(RNA)
+def _genbank_to_rna(fh, rec_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), rec_num)
+    return _construct(record, DNA, **kwargs).transcribe()
+
+
+@genbank.reader(Protein)
+def _genbank_to_protein(fh, rec_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), rec_num)
+    return _construct(record, Protein, **kwargs)
+
+
+@genbank.writer(None)
+def _sequences_to_genbank(fh):
+    pass
+
+
+def _construct(record, constructor=None, **kwargs):
+    seq, md, pmd = record
+    if constructor is None:
+        unit = md['LOCUS']['unit']
+        if unit == 'bp':
+            # RNA mol type has T instead of U;
+            # so still read in as DNA
+            constructor_ = DNA
+        elif unit == 'aa':
+            constructor_ = Protein
+        else:
+            constructor_ = Sequence
+    else:
+        constructor_ = constructor
+    return constructor_(seq, metadata=md, positional_metadata=pmd, **kwargs)
+
+
+def _parse_genbanks(fh):
+    data_chunks = []
+    for line in _line_generator(fh, skip_blanks=True, strip=False):
+        if line.startswith('//'):
+            yield _parse_single_genbank(data_chunks)
+            data_chunks = []
+        else:
+            data_chunks.append(line)
+
+
+def _parse_single_genbank(chunks):
+    metadata = dict()
+    metadata['REFERENCE'] = []
+    sequence = ''
+    # each section starts with a HEADER without indent.
+    section_splitter = yield_section(lambda x: not x[0].isspace(), strip=False)
+    for section in section_splitter(chunks):
+        header = section[0].split(None, 1)[0]
+        parser = globals().get('_parse_%s' % header.lower())
```

Comment:
This is pretty cool but I think an explicit lookup table will be safer. Whatever this module imports will be added to `globals()`, so it isn't guaranteed to only search for `_parse_*` functions defined in the file. This could lead to strange bugs or errors if the module imports a package-private function that happens to match the naming scheme (for example, importing `skbio.io.format._base._parse_fasta_like_header`).

Other reasons to use an explicit lookup table:

- there is reduced cognitive load for maintainers because they don't have to remember that there is a specific function naming scheme
- developers will have more freedom to name the parsing functions what they want